torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -0,0 +1,319 @@
+        *context, {reinterpret_cast<T*>(input_buffer->data()}, count, src_rank
+      ),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T> // unused
+struct algorithm_spec<DataOperation::BARRIER, T> {
+  static GlooCache::key_type key(THDGroup group_id) {
+    return std::make_tuple(DataOperation::BARRIER, group_id, 0, 0, 0);
+  }
+
+  static GlooCache::value_type create(const DataChannel::Group& group) {
+    auto context = GlooCache::get().createContext(group);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BarrierAllToAll>(*context),
+      nullptr,
+      nullptr
+    );
+  }
+};
+
+
+struct GlooCache {
+  using key_type = std::tuple<
+    DataOperation, THDGroup, std::size_t, std::size_t, THDReduceOp
+  >;
+  using value_type = std::tuple<
+    std::shared_ptr<context_type>,
+    std::shared_ptr<algorithm_type>,
+    std::shared_ptr<buffer_type>,
+    std::shared_ptr<buffer_type>
+  >;
+
+  using ret_type = std::tuple<
+    std::shared_ptr<algorithm_type>, // algorithm
+    void*,                           // input buffer (nullptr if not used)
+    void*                            // output buffer (nullptr if not used)
+  >;
+
+  GlooCache(GlooCache const&)       = delete;
+  void operator=(GlooCache const&)  = delete;
+
+  // singleton instance
+  static GlooCache& get() {
+    static GlooCache instance;
+    return instance;
+  }
+
+  // TODO: enforce setters to be the first thing called (maybe some kind of factory)
+  void setStoreGetter(std::function<PrefixStore ()> store_getter) {
+    _store_getter = store_getter;
+  }
+
+  void setRank(rank_type rank) {
+    _rank = rank;
+  }
+
+  std::make_shared<context_type> createContext(const DataChannel::Group& group) {
+    // For now we have to get store even we are not in group to increase static `id` prefix
+    // See `getStore()` function implementation in `DataChannelGloo`.
+    auto store = _store_getter();
+
+    rank_type group_rank;
+    bool exists;
+    std::tie(group_rank, exists) = group.getGroupRank(_rank);
+    if (!exists)
+      continue;
+
+    auto context = std::make_shared<context_type>(group_rank, group.size());
+    context->connectFullMesh(store, _device);
+    return context;
+  }
+
+  std::make_shared<buffer_type> createBuffer(std::size_t bytes) {
+    return std::make_shared<buffer_type>(buffer_type(bytes));
+  }
+
+  template<DataOperation D, typename T, typename... Args>
+  ret_type getAlgorithm(THDGroup group_id, const DataChannel::Group& group,
+                        Args&&... args) {
+    auto key = algorithm_spec<D, T>::key(group_id, std::forward<Args>(args)...);
+    if (_algorithms.find(key) == _algorithms.end()) {
+      _algorithms[key] = std::move(algorithm_spec<D, T>::create(group, std::forward<Args>(args)...));
+    }
+
+    auto& value = _algorithms[key];
+    return std::make_tuple(
+      std::get<1>(value),
+      std::get<2>(value)->data(),
```

Comment:
Hmm, I think no. We save buffers (currently vectors) in `_algorithms` variable and in this line we get raw pointer to those buffers. So I untill those buffers are not destroyed those raw pointers will be still valid, right? And since `GlooCache` has static singleton instance they should live until the program ends.